### PR TITLE
Support RPN function validator

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR adds validator to RPN function, as well as its codegen. Validators can verify whether input expression node is valid, for example, with correct data types and number of arguments.

- [x] Merge https://github.com/tikv/tikv/pull/5020 first.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

New unit tests.

### CODE REVIEW

File: components/cop_codegen/src/rpn_function.rs [hunk 1]
```diff
@@ -9,15 +11,31 @@ use syn::*;
...
     is_varg: bool,
+
+    /// Whether or not the function is a raw varg function. Raw varg function accepts `&[ScalarValueRef]`.
     is_raw_varg: bool,
+
+    /// The minimal accepted arguments, which will be checked in validator.
+    ///
+    /// Only varg or raw varg function accepts a range of number of arguments. Other kind of
+    /// function strictly stipulates number of arguments according to the function definition.
+    min_args: Option<usize>, // Only for varg
```
Comment (sticnarf) [inline components/cop_codegen/src/rpn_function.rs]: "Only for varg or raw varg" for clarity

File: components/cop_codegen/src/rpn_function.rs [hunk 2]
```diff
@@ -573,6 +745,26 @@ impl NormalRpnFn {
...
+            );
+            validator_fn_ptr = quote! { validate #ty_generics_turbofish };
+        } else {
+            validator_fn = None;
+            validator_fn_ptr = self
+                .validator
+                .as_ref()
+                .map(|f| f.into_token_stream())
+                .unwrap();
+        }
```
Comment (sticnarf) [inline components/cop_codegen/src/rpn_function.rs]: This part of code looks similar among the three types of RPN functions. Can we extract it out?

File: src/coprocessor/dag/rpn_expr/types/expr_builder.rs [hunk 3]
```diff
@@ -301,8 +298,14 @@ where
     let func_meta = fn_mapper(tree_node.get_sig(), tree_node.get_children())?;
 
     // Validate the input expression.
-    // (func_meta.validator_ptr)(&tree_node)
-    //     .map_err(|e| Error::Other(box_err!("Invalid function signature: {}", e)))?;
+    (func_meta.validator_ptr)(&tree_node).map_err(|e| {
+        Error::Other(box_err!(
+            "Invalid {} (sig = {:?}) signature: {}",
+            func_meta.name,
+            tree_node.get_tp(),
```
Comment (sticnarf) [inline src/coprocessor/dag/rpn_expr/types/expr_builder.rs]: `tree_node.get_sig()`?

File: components/cop_codegen/src/rpn_function.rs [hunk 4]
```diff
@@ -88,6 +129,76 @@ mod kw {
...
+            self.tokens.push(quote! {
+                function::validate_expr_arguments_gte(expr, #min_args)?;
+            });
+        }
+        self
+    }
+
+    pub fn validate_args_identical_type(mut self, args_evaluable: &TypePath) -> Self {
+        self.tokens.push(quote! {
+            for children in expr.get_children() {
```
Comment (sticnarf) [inline components/cop_codegen/src/rpn_function.rs]: `children` -> `child` :laughing:

File: components/cop_codegen/src/rpn_function.rs [hunk 5]
```diff
@@ -75,10 +102,24 @@ impl parse::Parse for RpnFnAttr {
                 "`varg` and `raw_varg` conflicts to each other",
             ));
         }
+        if !is_varg && !is_raw_varg && min_args != None {
+            return Err(Error::new_spanned(
+                config_items,
+                "`min_args` is only available when `varg` or `raw_varg` presents",
+            ));
+        }
+        if validator.is_some() && min_args != None {
```
Comment (lonng) [inline components/cop_codegen/src/rpn_function.rs]: I think the `min_args` and `validator` can coexist (Maybe we can append generated tokens to customized tokens?).

Comment (breeswish) [reply components/cop_codegen/src/rpn_function.rs]: Very good suggestion 👏

**Pull request merged**